test/benchmark_http_client_test.cc

```diff
@@ -41,41 +41,70 @@ class BenchmarkClientHttpTest : public Test {
 
     auto& tracer = static_cast<Envoy::Tracing::MockHttpTracer&>(*http_tracer_);
     EXPECT_CALL(tracer, startSpan_(_, _, _, _))
-        .WillRepeatedly(
-            Invoke([&](const Envoy::Tracing::Config& config, const Envoy::Http::HeaderMap&,
-                       const Envoy::StreamInfo::StreamInfo&,
-                       const Envoy::Tracing::Decision) -> Envoy::Tracing::Span* {
-              EXPECT_EQ(Envoy::Tracing::OperationName::Egress, config.operationName());
-              auto* span = new NiceMock<Envoy::Tracing::MockSpan>();
-              return span;
-            }));
-    request_generator_ = []() {
+        .WillRepeatedly([&](const Envoy::Tracing::Config& config, const Envoy::Http::HeaderMap&,
+                            const Envoy::StreamInfo::StreamInfo&,
+                            const Envoy::Tracing::Decision) -> Envoy::Tracing::Span* {
+          EXPECT_EQ(Envoy::Tracing::OperationName::Egress, config.operationName());
+          auto* span = new NiceMock<Envoy::Tracing::MockSpan>();
+          return span;
+        });
+  }
+  //Default function for request generator when the content doesn't matter.
+  RequestGenerator getDefaultRequestGenerator() const {
+    RequestGenerator request_generator = []() {
       auto header = std::make_shared<Envoy::Http::TestRequestHeaderMapImpl>(
           std::initializer_list<std::pair<std::string, std::string>>(
               {{":scheme", "http"}, {":method", "GET"}, {":path", "/"}, {":host", "localhost"}}));
       return std::make_unique<RequestImpl>(header);
     };
+    return request_generator;
   }
-
-  void testBasicFunctionality(const uint64_t max_pending, const uint64_t connection_limit,
-                              const uint64_t amount_of_request) {
+  /// Helper function to get properties in a map that should be verified during the test
+  absl::flat_hash_map<std::string, std::string>
+  getTestRecordedProperties(const Envoy::Http::RequestHeaderMap& header) {
+    absl::flat_hash_map<std::string, std::string> properties_map;
+    properties_map["uri"] = std::string(header.getPathValue());
+    return properties_map;
+  }
+  /// Used to set up benchmarkclient. Especially from within the testBenchmarkClientFunctionality.
+  void setupBenchmarkClient(const RequestGenerator& request_generator) {
+    client_ = std::make_unique<Client::BenchmarkClientHttpImpl>(
+        *api_, *dispatcher_, store_, std::make_unique<StreamingStatistic>(),
+        std::make_unique<StreamingStatistic>(), std::make_unique<StreamingStatistic>(),
+        std::make_unique<StreamingStatistic>(), false, cluster_manager_, http_tracer_, "benchmark",
+        request_generator, true);
+  }
+  /// Primary testing method. Confirms that connection limits are met and number of requests are
+  /// correct. If specified, also checks the header expectations, if not specified, it is ignored.
```

Comment:
nit: I think you mean if empty. it's always specified as this is a const reference.

nit: We could add another boolean that controls the header validation instead of relying on the empty value.

Reply:
First one is done. Is it better to have a separate boolean? I vaguely feel like passing more parameters like that means I'm doing something wrong, but I'm not sure what the correct approach is.